internal/cutil/cslice.go

```diff
@@ -0,0 +1,92 @@
+package cutil
+
+// #include <stdlib.h>
+import "C"
+import "unsafe"
+
+// The following code needs some explanation:
+// This creates slices on top of the C memory buffers allocated before in
+// order to safely and comfortably use them as arrays. First the void pointer
+// is cast to a pointer to an array of the type that will be stored in the
+// array. Because the size of an array is a constant, but the real array size
+// is dynamic, we just use the biggest possible index value maxIdx, to make
+// sure it's always big enough. (Nothing is allocated by casting, so the size
+// can be arbitrarily big.) So, if the array should store items of myType, the
+// cast would be (*[maxIdx]myItem)(myCMemPtr).
+// From that array pointer a slice is created with the [start:end:capacity]
+// syntax. The capacity must be set explicitly here, because by default it
+// would be set to the size of the original array, which is maxIdx, which
+// doesn't reflect reality in this case. This results in definitions like:
+// cSlice := (*[maxIdx]myItem)(myCMemPtr)[:numOfItems:numOfItems]
+
+const maxIdx = 1<<31 - 1 // 2GB, max int32 value, should be safe
+const ptrSize = C.size_t(unsafe.Sizeof(unsafe.Pointer(nil)))
+
+// CPtr is a pointer to C memory.
+// Not required, but makes it more obvious, when only pointers to C memory and
+// no pointers to Go memory are allowed to be stored.
+type CPtr unsafe.Pointer
+
+// CSize is a C.size_t
+// (required, because C.size_t is different for every package)
+type CSize C.size_t
+
+////////// CPtr //////////
+
+// CPtrSlice is a C allocated slice of pointers.
+type CPtrSlice []CPtr
+
+// NewCPtrSlice returns a CPtrSlice.
+// Similar to CString it must be freed with Free(slice)
+func NewCPtrSlice(size int) CPtrSlice {
+	if size == 0 {
+		return nil
+	}
+	cMem := C.malloc(C.size_t(size) * ptrSize)
+	cSlice := (*[maxIdx]CPtr)(cMem)[:size:size]
+	return cSlice
+}
+
+// Ptr returns a pointer to CPtrSlice
+func (v *CPtrSlice) Ptr() CPtr {
+	if len(*v) == 0 {
+		return nil
+	}
+	return CPtr(&(*v)[0])
+}
+
+// Free frees a CPtrSlice
+func (v *CPtrSlice) Free() {
+	C.free(unsafe.Pointer(v.Ptr()))
+	*v = nil
```

Comment:
Side effecting the value so that its no longer a "real" value this way gives me a really bad feeling. I'd drop this.

Reply:
You are not allowed to use `v` afterwards anyway. So if a side effect on `v` gives you a bad feeling, it's because you were tempted to do something forbidden, which is exactly the reason I did this. Better having a `nil` slice, than a slice that might cause a segfault.

Reply:
By the way, that is not called side-effect, it's a non-constant method, and very common in OOP.

Reply:
Just for reference how you empty a vector (the C++ equivalent of a slice): http://www.cplusplus.com/reference/vector/vector/clear/

Reply:
BTW, I _have_ to do that to protect against double `Free()`